server/server.js

```diff
@@ -2,29 +2,78 @@ const express = require('express');
 const path = require('path');
 const bodyParser = require('body-parser');
 const cookieParser = require('cookie-parser');
-const quizerController = require('./controllers/quizController')
+
+const quizController = require('./controllers/quizController');
+const userController = require('./controllers/userController');
+const cookieController = require('./controllers/cookieController');
+const sessionController = require('./controllers/sessionController');
 
 const app = express();
+
+app.use(bodyParser.urlencoded({ extended: true }));
 app.use(express.json());
+app.use(cookieParser());
+
+app.use('/build', express.static(path.resolve(__dirname, '../build')));
+
+app.get('/', (req, res) => {
+  //get request to login/signup page
+  return res.status(200).sendFile(path.resolve(__dirname, '../index.html'));
+});
+
+app.post(
+  '/signup',
+  userController.createUser,
+  cookieController.setSSIDCookie,
+  sessionController.startSession,
+  (req, res) => {
+    // on failed signup, send boolean false
+    if (res.locals.alreadyExists)
+      return res
+        .status(200)
+        .json({ message: 'Username already taken!', loggedIn: false });
+    // on successful signup, send boolean true
+    res.status(200).json({ message: 'New user added!', loggedIn: true });
+  }
+);
 
-app.use('/build', express.static(path.resolve(__dirname, '../build')))
+app.post(
+  '/login',
+  userController.verifyUser,
+  cookieController.setSSIDCookie,
+  sessionController.startSession,
+  (req, res) => {
+    if (!res.locals.loggedIn)
+      // on failed sign in, send boolean false- to update
+      return res
+        .status(200)
+        .json({ message: 'Incorrect username/password', loggedIn: false });
+    // on successful sign in, send boolean true - to update
+    res.status(200).json({ message: 'Log in successful', loggedIn: true });
+  }
+);
 
-app.get('/', (req,res)=> {
-    //get request to login/ signup page
-    return res.status(200).
-    sendFile(path.resolve(__dirname,'../index.html'))
-})
+app.get(
+  '/quiz-overflow',
+  sessionController.isLoggedIn,
+  quizController.getQuestion,
+  (req, res) => {
+    console.log('session cookieSessionMatch', res.locals.cookieSessionMatch);
+    // after frontend is ready to test, see if we can redirect to '/' in the case a session expires
+    // after logging in or if we need to send a res.locals with empty key values for question and choices.
+    if (!res.locals.cookieSessionMatch) res.status(200).json('Invalid session');
+    res.status(200).json(res.locals);
+  }
+);
 
-app.get('/quizoverflow', quizerController.getQuestion, (req,res)=>{
-    res.status(200)
-    res.json(res.locals)
-})
+app.use((req, res, next) => {
+  res.status(404).send('Not Found');
+});
 
-app.use((req, res, next)=>{
-    res.status(404)
-    .send('Not Found')
-}
-)
+app.use((err, req, res, next) => {
+  console.log('error handler', err);
+  res.status(500).send('Internal Server Error');
```

Comment:
I was told in my APC to always start this with `return` as best practice.

Reply:
Agreed, updated.

Reply:
Looks good.